server.go

```diff
@@ -97,16 +93,44 @@ func (s *Server) ServeHTTPS(addr, keyfile, certfile string, ready *chan bool) er
 	if s.listener, err = listenTLS(addr, keyfile, certfile); err != nil {
 		return err
 	}
+	s.tls = true
+	fmt.Printf("Listen http on %s\n", addr)
+	return s.doServe(ready)
+}
+
+func (s *Server) doServe(ready *chan bool) error {
+	// A dirty trick to associate a connection with the http.Request it
+	// contains. In "net/http/server.go", handler will be called
+	// immediately after ConnState changed to StateActive, so it's safe to
+	// loop through all elements in a channel to find a match remote addr.
+	q := make(chan net.Conn, 10)
 
-	// Set up server
 	proxy := http.HandlerFunc(
 		func(w http.ResponseWriter, req *http.Request) {
+			for c := range q {
+				if c.RemoteAddr().String() == req.RemoteAddr {
```

Comment:
This line should work if supporting only IPv4. With IPv6, different ip strings could refer to the same IP. The problem is that there is no guarantee that both sides have been "normalized" to the same representation or that the request structure just copies the address from the lower network layers without any kind of potential transformation.
I think we are ok, but it is possible that we see some issues, particularly with IPv6.